Kube proxy node manager

aroradaman opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This carries the work done in https://github.com/kubernetes/kubernetes/pull/125382 consolidates and simplifies how kube-proxy deals and watches over Node objects.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aroradaman commented:
/test pull-kubernetes-unit
/test pull-kubernetes-typecheck
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-conformance-kind-ga-only-parallel

aroradaman commented:
/test pull-kubernetes-unit
/test pull-kubernetes-typecheck
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-conformance-kind-ga-only-parallel

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -620,7 +621,13 @@ func (s *ProxyServer) Run(ctx context.Context) error {
+	// or an independent binary, since it may run before the kubelet on the node updates the Node.
+	// The solution in the meantime is to process the Node events and crash if the NodeIPs
+	// has changed since the first read.
+	// https://issues.k8s.io/111321
```
This isn't about "stale data" in that sense; it's not that kube-proxy is seeing incorrect data, it's that it's possible for the data to change after kube-proxy first learns it (if the node is deleted and then recreated). Those features wouldn't help.

danwinship commented on pkg/proxy/node.go:
```diff
@@ -109,3 +111,55 @@ func (n *NodeEligibleHandler) OnNodeDelete(node *v1.Node) { n.HealthServer.SyncN
+
+// NodeManager handles the life cycle of kube-proxy based on the NodeIPs.
+// Implements the config.NodeHandler interface
+// (ref: https://issues.k8s.io/111321)
```
Don't link to an issue (that probably talks about lots of other things as well), just explain what it's doing and why.

danwinship commented on pkg/proxy/node.go:
```diff
@@ -109,3 +111,55 @@ func (n *NodeEligibleHandler) OnNodeDelete(node *v1.Node) { n.HealthServer.SyncN
+	n.OnNodeUpsert(node)
+}
+
+func (n *NodeManager) OnNodeUpsert(node *v1.Node) {
```
"upsert"? bleah. How about `updated`

danwinship commented on pkg/proxy/node.go:
```diff
@@ -109,3 +111,55 @@ func (n *NodeEligibleHandler) OnNodeDelete(node *v1.Node) { n.HealthServer.SyncN
+
+// OnNodeDelete is a handler for Node deletes.
+func (n *NodeManager) OnNodeDelete(node *v1.Node) {
+	n.logger.Error(nil, "Current Node is being deleted", "node", klog.KObj(node))
```
(by the end of the PR we should be exiting in this case too probably)

There's some discussion of what cases we need to care about in https://github.com/kubernetes/kubernetes/pull/125382#discussion_r1669405761

danwinship commented on pkg/proxy/node.go:
```diff
@@ -109,3 +111,55 @@ func (n *NodeEligibleHandler) OnNodeDelete(node *v1.Node) { n.HealthServer.SyncN
+	n.logger.Error(nil, "Current Node is being deleted", "node", klog.KObj(node))
+}
+
+// OnNodeSynced is a handler for Node syncs.
```
`// OnNodeSynced is called after the cache is synced and all pre-existing Nodes have been reported`

danwinship commented on pkg/proxy/node_test.go:
```diff
@@ -146,6 +151,70 @@ func TestNodePodCIDRHandlerUpdate(t *testing.T) {
+	defer func() {
+		klog.OsExit = oldKlogOsExit
+	}()
+	klog.OsExit = customExit
```
uh.... I don't like that. How about having a `NodeManager.exit` that the unit tests can override instead?

oh, lol, you're just copying existing tests. So that makes it _definitely_ not-parallel-execution-safe.

danwinship commented on cmd/kube-proxy/app/server_linux.go:
```diff
@@ -381,48 +376,39 @@ func getConntrackMax(ctx context.Context, config proxyconfigapi.KubeProxyConntra
+	var node *v1.Node
+	var err error
+
+	pollErr := wait.PollUntilContextCancel(ctx, pollInterval, true, func(context.Context) (bool, error) {
```
We don't need two very similar methods (`getPodIPs` and `getPodCIDRs`). We can just have one loop that you can tell it to wait for either just node IPs, or both node IPs and pod CIDRs. And the code could be in `NodeManager` rather than in `cmd/kube-proxy`.

danwinship commented on pkg/proxy/node.go:
```diff
@@ -112,18 +48,22 @@ func (n *NodeEligibleHandler) OnNodeDelete(node *v1.Node) { n.HealthServer.SyncN
 }
 
-func NewNodeManager(ctx context.Context, nodeIPs []net.IP) *NodeManager {
+func NewNodeManager(ctx context.Context, nodeIPs []net.IP, podCIDRs []string, localModeNodeCIDR bool) *NodeManager {
```
maybe `watchPodCIDRs` instead of `localModeNodeCIDR`?

danwinship commented on pkg/proxy/node.go:
```diff
@@ -28,41 +28,23 @@ import (
 }
 
-func NewNodeManager(ctx context.Context, nodeIPs []net.IP, podCIDRs []string, localModeNodeCIDR bool) *NodeManager {
+func NewNodeManager(ctx context.Context, nodeIPs []net.IP, podCIDRs []string, localModeNodeCIDR bool, healthServer *healthcheck.ProxyHealthServer) *NodeManager {
```
When I was thinking about this, I wasn't sure if it would make more sense for `NodeManager` to take a `ProxyHealthServer` argument, or for `ProxyHealthServer` to take a `NodeManager` argument.

In particular, note that `ProxyHealthServer` doesn't actually react immediately to `SyncNode`; it doesn't actually look at the node object until the next time it constructs a health check response, so in theory, it could just call `nodeManager.GetCurrentNode()` at that point.

(But again, I didn't actually decide for sure that this made sense, I just wondered if it did.)

danwinship commented on pkg/proxy/config/config.go:
```diff
@@ -259,12 +259,8 @@ func (c *ServiceConfig) handleDeleteService(obj interface{}) {
-	// node object is observed.
-	OnNodeUpdate(oldNode, node *v1.Node)
+	// OnNodeUpsert is called whenever a node object is created or updated.
+	OnNodeUpsert(node *v1.Node)
```
(still don't like the name)

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -612,34 +612,8 @@ func (proxier *Proxier) OnEndpointSlicesSynced() {
-// node object is observed.
-func (proxier *Proxier) OnNodeUpdate(oldNode, node *v1.Node) {
+// OnNodeUpsert is called whenever creation of new or modification of an existing node object is observed.
+func (proxier *Proxier) OnNodeUpsert(node *v1.Node) {
```
Eventually, the Proxier-level signal should just be `OnTopologyChange(topologyLabels map[string]string)`, the same way that for ServiceCIDR we didn't expose Add/Update/Delete events, we just have a single `OnServiceCIDRsChanged(cidrs []string)`.

(One way to do that would be to have `config.NodeConfig` for `NodeManager` and a separate `config.NodeTopologyConfig` interface for `proxy.Provider`. Another way would be to not have the `proxy.Provider` watch nodes directly, and instead have it just receive updates from the `NodeManager`, somehow?)

danwinship reviewed (commented):
I know this is WIP, but some initial comments...

aroradaman commented:
/cc @danwinship @aojea @adrianmoisey

aroradaman commented:
/retest

adrianmoisey commented on pkg/proxy/node.go:
```diff
@@ -18,94 +18,152 @@ package proxy
+// NewNodeManager returns NodeManager after waiting for the node object to exist and have NodeIPs
+// and PodCIDRS (if watchPodCIDRs is true)
+func NewNodeManager(ctx context.Context, nodeLister corelisters.NodeLister, nodeName string, watchPodCIDRs bool) (*NodeManager, error) {
+	// we wait for at most 5 minutes for allocators to assign a PodCIDR to the node after it is registered.
```
I don't know any of the background here, is there a reason for 5 minutes? Based on this PR I assume it's enough time, but I'm wondering if there are cases where there's need for this to be longer (or configurable by the user?)

aroradaman replied:
I don';t think it will be a good idea to allow users to configure such a granular configuration, this may lead to a pattern to allow configurations for everything. 

https://github.com/kubernetes/kubernetes/blob/627ce4946ccd6b67bc18f5a4dda92cf6c583e9b7/cmd/kube-proxy/app/server_linux.go#L56-L58

aroradaman committed:
kube-proxy: add NodeTopologyConfig for tracking topology labels

This simplifies how the proxier receives update for change in node
labels. Instead of passing the complete Node object we just pass
the proxy relevant topology labels extracted from the complete list
of labels, and the downstream event handlers will only be notified
when there are changes in topology labels.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
kube-proxy: add NodeManager to manage life-cycle based on NodeIPs

NodeManager initialises node informers, waits for cache sync and polls for
node object to retrieve NodeIPs, handle node events and crashes kube-proxy
when change in NodeIPs is detected.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
kube-proxy: merge NodePodCIDRHandler with NodeManager

NodeManager, if configured with to watch for PodCIDR watch, watches
for changes in PodCIDRs and crashes kube-proxy if a change is
detected in PodCIDRs.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
kube-proxy: merge NodeEligibleHandler with NodeManager

ProxyHealthServer now consumes NodeManager to get the latest
updated node object for determining node eligibility.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
kube-proxy: merge OnNodeAdd and OnNodeUpdate into OnNodeChange

For kube-proxy, node addition and node update is semantically
considered as similar event, we have exactly same handler
logic for these two events resulting in duplicate code and
unit tests.
This merges the `NodeHandler` interface methods OnNodeAdd and
OnNodeUpdate into OnNodeChange along with the implementation
of the interface.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman commented on pkg/proxy/node.go:
```diff
@@ -146,28 +88,52 @@ func newNodeManager(ctx context.Context, nodeLister corelisters.NodeLister, node
+
+	// we return error is watchPodCIDRs was configured and PodCIDRs were not retrieved.
+	if watchPodCIDRs && len(podCIDRs) == 0 {
+		return nil, err
```
This will handle case when PodCIDRs were detected but NodeIPs were not. Check on error will return the error, but we are silently ignoring if not able to detect NodeIP.

aroradaman commented on pkg/proxy/topology.go:
```diff
@@ -191,15 +191,3 @@ func filterEndpoints(endpoints []Endpoint, predicate func(Endpoint) bool) []Endp
 }
-
-// ExtractTopologyLabels consumes node labels and extract topology related labels if present.
-func ExtractTopologyLabels(nodeLabels map[string]string) map[string]string {
```
Yes, but only notifying when there is change in proxy related labels made more sense to me. How about https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/apis/well_known_labels.go for a common place?

aroradaman commented:
/cc @aojea

danwinship commented on pkg/proxy/node.go:
```diff
@@ -109,3 +122,111 @@ func (n *NodeEligibleHandler) OnNodeDelete(node *v1.Node) { n.HealthServer.SyncN
+			return false, nil
+		}
+		return true, nil
+	})
```
you're discarding the error?

It seems to me like if kube-proxy can't get the node IPs, then it ought to fail... Can you do look through the old PRs that resulted in the current code to figure out why we made it keep going?

aroradaman replied:
We do fallback to localhost (IPv4 and IPv6) when NodeIPs are not detected
https://github.com/kubernetes/kubernetes/blob/b98b86bd6abb24e3d2451a885354f09cd7af22a2/cmd/kube-proxy/app/server.go#L636-L649

I'll go through the old PRs.

aroradaman replied:
Historically `--bind-address` was directly used as NodeIP for primary IP Family detection, and https://github.com/kubernetes/kubernetes/commit/56df70b639c5b5bf53076a3483a2e40b8f60c49a (https://github.com/kubernetes/kubernetes/pull/91725) added `detectNodeIP` which fetched NodeIP from Node object if the `--bind-address` was unspecified and fall backed to localhost on failure.

https://github.com/kubernetes/kubernetes/commit/0df4a69f5c90990d47d1337e5fc99a764ff99474 added the polling to wait for NodeIPs to be available with Node object, but never returned any error on poll timeout.

danwinship replied:
So, I guess the reason we ignore errors is because the node IP is _mostly_ not actually used, other than for deciding the primary IP family, so it seemed weird to make a `GetNodeHostIPs` failure be fatal... But at the same time, it really _should_ be fatal...

Either way, I dislike the fact that we completely throw the error away within NodeManager. If we're going to ignore the error, it ought to be done at the `cmd/kube-proxy` level.

And actually, the current code doesn't work anyway; kubelet always sets _some_ node IPs on the node when it first creates the Node object, so the only way we could get to the end of `newNodeManager` without having found node IPs would be if we didn't even find a Node, in which case we'd end up setting `n.node` to `nil` and then `n.NodeIPs()` would crash inside `utilnode.GetNodeHostIPs()`.

At this point we have basically normalized the fact that kube-proxy will exit and expect to be restarted in some circumstances. We should just treat the lack of a Node object as a fatal error and exit.

aroradaman commented on pkg/proxy/config/config_test.go:
```diff
@@ -457,3 +461,139 @@ func TestNewEndpointsMultipleHandlersAddRemoveSetAndNotified(t *testing.T) {
+				invoked.Store(false)
+				return nil
+			}
+			time.Sleep(time.Microsecond)
```
lol yes !

aroradaman commented:
> Hm... maybe rather than having it have to be in the same file as CategorizeEndpoints, you could just add a comment to CategorizeEndpoints pointing out that if you change the code to look at any other endpoints, you need to change ExtractTopologyLabels (wherever that is) as well...

> (Also, it would be best to have that code end up in the right place in the initial commit, rather than being added in one commit and then moved in the next commit. Or maybe that means swapping the order of the two commits?)

@danwinship I have squashed the two commits together, and added a comment in `CategorizeEndpoints`.

danwinship commented on pkg/proxy/healthcheck/proxy_health.go:
```diff
@@ -70,34 +71,31 @@ type ProxyHealthServer struct {
+		clock:                  c,
+		addr:                   addr,
+		healthTimeout:          healthTimeout,
+		nodeManager:            nodeManager,
```
there was a blank line before `lastUpdatedMap` before and if you leave that there you won't have to reindent everything

danwinship reviewed (commented):
sorry for the delay. very close now.

aroradaman commented on pkg/proxy/node.go:
```diff
@@ -208,3 +191,10 @@ func (n *NodeManager) OnNodeDelete(node *v1.Node) {
 // OnNodeSynced is called after the cache is synced and all pre-existing Nodes have been reported
 func (n *NodeManager) OnNodeSynced() {}
+
+// Node returns the latest copy of node object.
```
Is it fine if I skip this comment and return a deep copy instead?

aroradaman commented on cmd/kube-proxy/app/server.go:
```diff
@@ -207,7 +207,14 @@ func newProxyServer(ctx context.Context, config *kubeproxyconfig.KubeProxyConfig
+	}
+
+	rawNodeIPs := s.NodeManager.NodeIPs()
+	logger.Info("Successfully retrieved NodeIPs", "NodeIPs", rawNodeIPs)
```
> rawNodeIPs can be nil/empty here if it timed out, right? We need to log an error somewhere if that happens;

rawNodeIPs will never be nil, assuming [kubelet always sets some node IPs on the node when it first creates the Node object](https://github.com/kubernetes/kubernetes/pull/130837#discussion_r2110059585).
If the node manager times out waiting for existence of node object then it will throw an error that will cause kube-proxy to fail.

In the current code `s.NodeManager.NodeIPs()` will never be called when node manager times out.

danwinship commented:
you dropped half of the branch on re-push...

aroradaman commented:
/test pull-kubernetes-e2e-gce
```
ctr: failed to dial "/run/containerd/containerd.sock": context deadline exceeded: connection error: desc = "transport: error while dialing: dial unix:///run/containerd/containerd.sock: timeout"
```

danwinship reviewed (commented):
ok, just a few comments on the unit tests that I hadn't looked too deeply at before, and a followup to my previous review comments on deepcopying...

aroradaman commented:
/release-note-none
